src/devices/Arduino/ArduinoBoard.cs

```diff
@@ -0,0 +1,514 @@
+        /// <returns>True on success, false if no board was found</returns>
+        public static bool TryFindBoard(IEnumerable<string> comPorts, IEnumerable<int> baudRates,
+#if !NETCOREAPP2_1
+            [NotNullWhen(true)]
+#endif
+            out ArduinoBoard? board)
+        {
+            foreach (var port in comPorts)
+            {
+                foreach (var baud in baudRates)
+                {
+                    ArduinoBoard? b = null;
+                    try
+                    {
+                        b = new ArduinoBoard(port, baud);
+                        b.Initialize();
+                        board = b;
+                        return true;
+                    }
+                    catch (Exception x) when (x is NotSupportedException || x is TimeoutException || x is IOException || x is UnauthorizedAccessException)
+                    {
+                        b?.Dispose();
+                    }
+                }
+            }
+
+            board = null!;
+            return false;
+        }
+
+        /// <summary>
+        /// Searches all available com ports for an Arduino device.
+        /// </summary>
+        /// <param name="board">A board, already open and initialized. Null if none was found.</param>
+        /// <returns>True if a board was found, false otherwise</returns>
+        /// <remarks>
+        /// Scanning serial ports may affect unrelated devices. If there are problems, use the
+        /// <see cref="TryFindBoard(System.Collections.Generic.IEnumerable{string},System.Collections.Generic.IEnumerable{int},out Iot.Device.Arduino.ArduinoBoard?)"/> overload excluding ports that shall not be tested.
+        /// </remarks>
+        public static bool TryFindBoard(
+#if !NETCOREAPP2_1
+            [NotNullWhen(true)]
+#endif
+            out ArduinoBoard? board)
+        {
+            return TryFindBoard(SerialPort.GetPortNames(), CommonBaudRates(), out board);
+        }
+
+        /// <summary>
+        /// Returns a list of commonly used baud rates.
+        /// </summary>
+        public static List<int> CommonBaudRates()
+        {
+            return new List<int>()
+            {
+                9600,
+                19200,
+                18400,
+                57600,
+                115200,
+                230400,
+                250000,
+                500000,
+                1000000,
+                2000000,
+            };
+        }
+
+        /// <summary>
+        /// Initialize the board connection. This must be called before any other methods of this class.
+        /// </summary>
+        /// <exception cref="NotSupportedException">The Firmata firmware on the connected board is too old.</exception>
+        /// <exception cref="TimeoutException">There was no answer from the board</exception>
+        public virtual void Initialize()
+        {
+            lock (_initializationLock)
+            {
+                if (_initialized)
```

Comment:
I'd recommend copying this if just before the lock as well and remove all `if (!_initialized) Initialize();` from the code

Reply:
Yea, that was needlessly complicated.